crates/uv/src/commands/tool/mod.rs

```diff
@@ -6,3 +13,71 @@ pub(crate) mod run;
 pub(crate) mod uninstall;
 pub(crate) mod update_shell;
 pub(crate) mod upgrade;
+
+#[derive(Debug, Clone)]
+pub(crate) enum Target<'a> {
+    /// e.g., `ruff`
+    Unspecified(&'a str),
+    /// e.g., `ruff@0.6.0`
+    Version(&'a str, Version),
+    /// e.g., `ruff@latest`
+    Latest(&'a str),
+    /// e.g., `--from ruff==0.6.0`
+    UserDefined(&'a str, &'a str),
+}
+
+impl<'a> Target<'a> {
+    /// Parse a target into a command name and a requirement.
+    pub(crate) fn parse(target: &'a str, from: Option<&'a str>) -> Self {
+        if let Some(from) = from {
+            return Self::UserDefined(target, from);
+        }
+
+        // e.g. `ruff`, no special handling
+        let Some((name, version)) = target.split_once('@') else {
+            return Self::Unspecified(target);
+        };
+
+        // e.g. `ruff@`, warn and treat the whole thing as the command
+        if version.is_empty() {
+            debug!("Ignoring empty version request in command");
+            return Self::Unspecified(target);
+        }
+
+        // e.g., ignore `git+https://github.com/astral-sh/ruff.git@main`
+        if PackageName::from_str(name).is_err() {
+            debug!("Ignoring non-package name `{name}` in command");
+            return Self::Unspecified(target);
+        }
+
+        match version {
+            // e.g., `ruff@latest`
+            "latest" => return Self::Latest(name),
+            // e.g., `ruff@0.6.0`
+            version => {
+                if let Ok(version) = Version::from_str(version) {
+                    return Self::Version(name, version);
+                }
+            }
+        };
+
+        // e.g. `ruff@invalid`, warn and treat the whole thing as the command
+        debug!("Ignoring invalid version request `{version}` in command");
+        Self::Unspecified(target)
+    }
+
+    /// Returns the name of the executable.
+    pub(crate) fn executable(&self) -> &str {
+        match self {
+            Self::Unspecified(name) => name,
+            Self::Version(name, _) => name,
+            Self::Latest(name) => name,
+            Self::UserDefined(name, _) => name,
+        }
```

Comment:
Does this collide with #6683?

Reply:
I was hoping to merge that today too. Unsure if it's 100% right even though Konsti tried it?